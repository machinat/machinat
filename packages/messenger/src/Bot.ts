import invariant from 'invariant';
import Engine, { DispatchError } from '@sociably/core/engine';
import Queue from '@sociably/core/queue';
import Renderer from '@sociably/core/renderer';
import ModuleUtilitiesI from '@sociably/core/base/ModuleUtilities';
import { makeClassProvider } from '@sociably/core/service';
import type {
  SociablyNode,
  SociablyBot,
  InitScopeFn,
  DispatchWrapper,
} from '@sociably/core';
import {
  MetaApiWorker,
  MetaApiJob,
  MetaApiResult,
  MetaApiDispatchResponse,
  MetaApiResponseBody,
} from '@sociably/meta-api';
import generalComponentDelegator from './components/general';
import { MESSENGER } from './constant';
import { ConfigsI, PlatformUtilitiesI } from './interface';
import MessengerChat from './Chat';
import SendTarget from './SendTarget';
import { createChatJobs, createAttachmentJobs } from './job';
import type {
  MessengerChannel,
  MessengerTarget,
  MessengerComponent,
  MessengerSegmentValue,
  MessengerDispatchFrame,
  MessengerSendOptions,
} from './types';

type MessengerBotOptions = {
  initScope?: InitScopeFn;
  dispatchWrapper?: DispatchWrapper<
    MetaApiJob,
    MessengerDispatchFrame,
    MetaApiResult
  >;
  pageId: string;
  accessToken: string;
  appSecret?: string;
  graphApiVersion?: string;
  apiBatchRequestInterval?: number;
};

/**
 * MessengerBot render messages and make API call to Messenger platform.
 * @category Provider
 */
export class MessengerBot
  implements SociablyBot<MessengerChannel, MetaApiJob, MetaApiResult>
{
  pageId: string;
  worker: MetaApiWorker;
  engine: Engine<
    MessengerChannel,
    MessengerSegmentValue,
    MessengerComponent<unknown>,
    MetaApiJob,
    MetaApiResult
  >;

  platform = MESSENGER;

  constructor({
    pageId,
    accessToken,
    appSecret,
    graphApiVersion = 'v11.0',
    apiBatchRequestInterval = 500,
    initScope,
    dispatchWrapper,
  }: MessengerBotOptions) {
    invariant(pageId, 'options.pageId should not be empty');
    this.pageId = pageId;

    invariant(accessToken, 'options.accessToken should not be empty');

    const renderer = new Renderer<
      MessengerSegmentValue,
      MessengerComponent<unknown>
    >(MESSENGER, generalComponentDelegator);

    const queue = new Queue<MetaApiJob, MetaApiResult>();
    const worker = new MetaApiWorker(
      accessToken,
      apiBatchRequestInterval,
      graphApiVersion,
      appSecret
    );

    this.engine = new Engine(
      MESSENGER,
      renderer,
      queue,
      worker,
      initScope,
      dispatchWrapper
    );
  }

  async start(): Promise<void> {
    this.engine.start();
  }

  async stop(): Promise<void> {
    this.engine.stop();
  }

  async render(
    target: string | MessengerTarget | MessengerChannel,
    messages: SociablyNode,
    options?: MessengerSendOptions
  ): Promise<null | MetaApiDispatchResponse> {
    const channel =
      typeof target === 'string'
        ? new MessengerChat(this.pageId, target)
        : target instanceof MessengerChat || target instanceof SendTarget
        ? target
        : new SendTarget(this.pageId, target);

    return this.engine.render(channel, messages, createChatJobs(options));
  }

  async renderAttachment(
    node: SociablyNode
  ): Promise<null | MetaApiDispatchResponse> {
    return this.engine.render(null, node, createAttachmentJobs);
  }

  async makeApiCall<ResBody extends MetaApiResponseBody>(
    method: 'GET' | 'PUT' | 'POST' | 'DELETE',
    relativeUrl: string,
    body: null | Record<string, unknown> = null
  ): Promise<ResBody> {
    try {
      const { results } = await this.engine.dispatchJobs(null, [
        {
          request: {
            method,
            relative_url: relativeUrl,
            body,
          },
        },
      ]);

      return results[0].body as ResBody;
    } catch (err) {
      if (err instanceof DispatchError) {
        throw err.errors[0];
      }
      throw err;
    }
  }
}

const BotP = makeClassProvider({
  lifetime: 'singleton',
  deps: [
    ConfigsI,
    { require: ModuleUtilitiesI, optional: true },
    { require: PlatformUtilitiesI, optional: true },
  ],
  factory: (
    { pageId, accessToken, appSecret, apiBatchRequestInterval },
    moduleUitils,
    platformUtils
  ) =>
    new MessengerBot({
      pageId,
      accessToken,
      appSecret,
      apiBatchRequestInterval,
      initScope: moduleUitils?.initScope,
      dispatchWrapper: platformUtils?.dispatchWrapper,
    }),
})(MessengerBot);

type BotP = MessengerBot;
export default BotP;
