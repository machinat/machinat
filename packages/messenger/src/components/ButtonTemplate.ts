import type {
  SociablyNode,
  NativeElement,
  AnyNativeComponent,
} from '@sociably/core';
import { formatNode } from '@sociably/core/utils';
import {
  makeUnitSegment,
  InnerRenderFn,
  UnitSegment,
} from '@sociably/core/renderer';
import { PATH_MESSAGES } from '../constant.js';
import type { MessageValue } from '../types.js';

/**
 * @category Props
 */
export type ButtonTemplateProps = {
  /**
   * Textual node with content up to 640 characters. Text will appear above the
   * buttons.
   */
  children: SociablyNode;
  /** 1-3 button elements to append after the text. */
  buttons: SociablyNode;
  sharable?: boolean;
};

export async function ButtonTemplate(
  node: NativeElement<ButtonTemplateProps, AnyNativeComponent>,
  path: string,
  render: InnerRenderFn
): Promise<UnitSegment<MessageValue>[]> {
  const { children, buttons, sharable } = node.props;
  const textSegments = await render(children, '.children');

  if (!textSegments) {
    throw new TypeError(`"children" prop should not be empty`);
  }
  for (const segment of textSegments) {
    if (segment.type !== 'text') {
      throw new TypeError(
        `non-textual node ${formatNode(
          segment.node
        )} received, only textual nodes allowed`
      );
    }
  }

  const buttonSegments = await render(buttons, '.buttons');
  const buttonValues = buttonSegments?.map((segment) => segment.value);

  return [
    makeUnitSegment(node, path, {
      type: 'message',
      apiPath: PATH_MESSAGES,
      params: {
        message: {
          attachment: {
            type: 'template',
            payload: {
              template_type: 'button',
              text: textSegments[0].value,
              sharable,
              buttons: buttonValues,
            },
          },
        },
      },
    }),
  ];
}
