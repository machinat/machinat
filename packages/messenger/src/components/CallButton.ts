import { NativeElement, AnyNativeComponent } from '@sociably/core';
import { makePartSegment, PartSegment } from '@sociably/core/renderer';

/**
 * @category Props
 */
export type CallButtonProps = {
  /** Button title, 20 character limit. */
  title: string;
  /**
   * Format must have "+" prefix followed by the country code, area code and
   * local number.
   */
  number: string;
};

export function CallButton(
  node: NativeElement<CallButtonProps, AnyNativeComponent>,
  path: string
): PartSegment<{}>[] {
  const { title, number } = node.props;
  return [
    makePartSegment(node, path, {
      type: 'phone_number',
      title,
      number,
    }),
  ];
}
