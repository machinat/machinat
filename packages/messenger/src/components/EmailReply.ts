/* eslint-disable import/prefer-default-export */
import { NativeElement, AnyNativeComponent } from '@sociably/core';
import { makePartSegment, PartSegment } from '@sociably/core/renderer';

const EMAIL_QUICK_REPLY_VALUES = { content_type: 'user_email' };

export function EmailReply(
  node: NativeElement<{}, AnyNativeComponent>,
  path: string
): PartSegment<{}>[] {
  return [makePartSegment(node, path, EMAIL_QUICK_REPLY_VALUES)];
}
