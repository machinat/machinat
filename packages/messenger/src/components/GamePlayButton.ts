import type { NativeElement, AnyNativeComponent } from '@sociably/core';
import { makePartSegment, PartSegment } from '@sociably/core/renderer';

/** @category Props */
export type GamePlayButtonProps = {
  /** Button title, 20 character limit. */
  title: string;
  /** This data will be sent to the game. */
  payload?: string;
  /** Player ID (Instant Game name-space) to play against. */
  playerId?: string;
  /** Context ID (Instant Game name-space) of the THREAD to play in. */
  contextId?: string;
};

export function GamePlayButton(
  node: NativeElement<GamePlayButtonProps, AnyNativeComponent>,
  path: string,
): PartSegment<{}>[] {
  const { title, payload, playerId, contextId } = node.props;
  return [
    makePartSegment(node, path, {
      type: 'game_play',
      title,
      payload,
      game_metadata:
        playerId || contextId
          ? {
              player_id: playerId,
              context_id: contextId,
            }
          : undefined,
    }),
  ];
}
