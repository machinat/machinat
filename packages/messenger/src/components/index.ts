export { default as renderGeneralComponents } from './general.js';
// media
export * from './Media.js';
// buttons
export * from './UrlButton.js';
export * from './PostbackButton.js';
export * from './CallButton.js';
export * from './LoginButton.js';
export * from './LogoutButton.js';
export * from './GamePlayButton.js';
// quick replies
export * from './TextReply.js';
export * from './EmailReply.js';
export * from './PhoneReply.js';
// handover protocols
export * from './PassThreadControl.js';
export * from './RequestThreadControl.js';
export * from './TakeThreadContorl.js';
// sender actions
export * from './MarkSeen.js';
export * from './TypingOn.js';
export * from './TypingOff.js';
// templates
export * from './GenericTemplate.js';
export * from './ButtonTemplate.js';
export * from './MediaTemplate.js';
export * from './ReceiptTemplate.js';
export * from './RequestOneTimeNotif.js';
// text
export * from './Latex.js';
// expression
export * from './Expression.js';
