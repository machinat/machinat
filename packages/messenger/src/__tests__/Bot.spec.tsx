import querystring from 'querystring';
import nock from 'nock';
import moxy, { Moxy } from '@moxyjs/moxy';
import Sociably from '@sociably/core';
import _Renderer from '@sociably/core/renderer';
import Queue from '@sociably/core/queue';
import _Engine from '@sociably/core/engine';
import { MetaApiWorker as _Worker, MetaApiError } from '@sociably/meta-api';
import { MessengerBot } from '../Bot';
import { Image, Expression, TextReply } from '../components';

const Renderer = _Renderer as Moxy<typeof _Renderer>;
const Engine = _Engine as Moxy<typeof _Engine>;
const Worker = _Worker as Moxy<typeof _Worker>;

nock.disableNetConnect();

jest.mock('@sociably/core/engine', () =>
  jest
    .requireActual('@moxyjs/moxy')
    .default(jest.requireActual('@sociably/core/engine'))
);

jest.mock('@sociably/core/renderer', () =>
  jest
    .requireActual('@moxyjs/moxy')
    .default(jest.requireActual('@sociably/core/renderer'))
);

jest.mock('@sociably/meta-api', () => {
  const module = jest.requireActual('@sociably/meta-api');
  return {
    ...module,
    MetaApiWorker: jest
      .requireActual('@moxyjs/moxy')
      .default(module.MetaApiWorker),
  };
});

const initScope = moxy(() => moxy());
const dispatchWrapper = moxy((x) => x);

const pageId = '1234567890';
const accessToken = '_ACCESS_TOKEN_';
const appSecret = '_APP_SECRET_';

const message = (
  <Expression quickReplies={<TextReply title="Hi!" payload="ðŸ‘‹" />}>
    Hello <b>World!</b>
    <Image url="https://sociably.io/greeting.png" />
  </Expression>
);

let graphApi;
const bodySpy = moxy(() => true);

beforeEach(() => {
  graphApi = nock('https://graph.facebook.com').post('/v11.0/', bodySpy);
  bodySpy.mock.clear();
  Engine.mock.clear();
  Renderer.mock.clear();
  Worker.mock.clear();
});

afterEach(() => {
  nock.cleanAll();
});

describe('#constructor(options)', () => {
  it('throw if accessToken not given', () => {
    expect(
      () => new MessengerBot({ pageId, appSecret } as never)
    ).toThrowErrorMatchingInlineSnapshot(
      `"options.accessToken should not be empty"`
    );
  });

  it('throw if pageId not given', () => {
    expect(
      () => new MessengerBot({ accessToken, appSecret } as never)
    ).toThrowErrorMatchingInlineSnapshot(
      `"options.pageId should not be empty"`
    );
  });

  it('assemble core modules', () => {
    const bot = new MessengerBot({
      initScope,
      dispatchWrapper,
      pageId,
      accessToken,
    });

    expect(bot.engine).toBeInstanceOf(Engine);

    expect(Renderer.mock).toHaveBeenCalledTimes(1);
    expect(Renderer.mock).toHaveBeenCalledWith(
      'messenger',
      expect.any(Function)
    );

    expect(Engine.mock).toHaveBeenCalledTimes(1);
    expect(Engine.mock).toHaveBeenCalledWith(
      'messenger',
      expect.any(Renderer),
      expect.any(Queue),
      expect.any(Worker),
      initScope,
      dispatchWrapper
    );

    expect(Worker.mock).toHaveBeenCalledTimes(1);
    expect(Worker.mock).toHaveBeenCalledWith(
      '_ACCESS_TOKEN_',
      500,
      'v11.0',
      undefined
    );
  });

  it('pass options to worker', () => {
    expect(
      new MessengerBot({
        initScope,
        dispatchWrapper,
        pageId,
        accessToken,
        appSecret,
        graphApiVersion: 'v8.0',
        apiBatchRequestInterval: 0,
      })
    );

    expect(Worker.mock).toHaveBeenCalledTimes(1);
    expect(Worker.mock).toHaveBeenCalledWith(
      '_ACCESS_TOKEN_',
      0,
      'v8.0',
      '_APP_SECRET_'
    );
  });
});

test('#start() and #stop() start/stop engine', () => {
  const bot = new MessengerBot({
    initScope,
    dispatchWrapper,
    pageId,
    accessToken,
    appSecret,
  });

  type MockEngine = Moxy<MessengerBot['engine']>;

  bot.start();
  expect((bot.engine as MockEngine).start.mock).toHaveBeenCalledTimes(1);

  bot.stop();
  expect((bot.engine as MockEngine).stop.mock).toHaveBeenCalledTimes(1);
});

describe('#render(channel, message, options)', () => {
  const bot = new MessengerBot({ pageId, accessToken, appSecret });

  let apiStatus;
  beforeEach(() => {
    const messageResult = {
      code: 200,
      body: JSON.stringify({ message_id: 'xxx', recipient_id: 'xxx' }),
    };
    apiStatus = graphApi.reply(200, [messageResult, messageResult]);
    bot.start();
  });

  afterEach(() => {
    bot.stop();
  });

  it('resolves null if message is empty', async () => {
    const empties = [undefined, null, [], <></>];
    for (const empty of empties) {
      // eslint-disable-next-line no-await-in-loop
      await expect(bot.render('john', empty)).resolves.toBe(null);
      expect(apiStatus.isDone()).toBe(false);
    }
  });

  it('send messages to me/messages api', async () => {
    const response = await bot.render('john', message);
    expect(response).toMatchSnapshot();

    for (const result of response!.results) {
      expect(result).toEqual({
        code: 200,
        body: { message_id: 'xxx', recipient_id: 'xxx' },
      });
    }

    expect(bodySpy.mock).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    for (const request of JSON.parse(body.batch)) {
      expect(request.method).toBe('POST');
      expect(request.relative_url).toBe('me/messages');
    }

    expect(body).toMatchSnapshot();
    expect(JSON.parse(body.batch).map((req) => querystring.decode(req.body)))
      .toMatchInlineSnapshot(`
      Array [
        Object {
          "message": "{\\"text\\":\\"Hello World!\\"}",
          "recipient": "{\\"id\\":\\"john\\"}",
        },
        Object {
          "message": "{\\"attachment\\":{\\"type\\":\\"image\\",\\"payload\\":{\\"url\\":\\"https://sociably.io/greeting.png\\"}},\\"quick_replies\\":[{\\"content_type\\":\\"text\\",\\"title\\":\\"Hi!\\",\\"payload\\":\\"ðŸ‘‹\\"}]}",
          "recipient": "{\\"id\\":\\"john\\"}",
        },
      ]
    `);

    expect(apiStatus.isDone()).toBe(true);
  });

  test('render options', async () => {
    const response = await bot.render('john', message, {
      messagingType: 'MESSAGE_TAG',
      tag: 'TRANSPORTATION_UPDATE',
      notificationType: 'SILENT_PUSH',
      personaId: 'billy17',
    });
    expect(response).toMatchSnapshot();

    for (const result of response!.results) {
      expect(result).toEqual({
        code: 200,
        body: { message_id: 'xxx', recipient_id: 'xxx' },
      });
    }

    expect(bodySpy.mock).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    expect(body).toMatchSnapshot();
    expect(JSON.parse(body.batch).map((req) => querystring.decode(req.body)))
      .toMatchInlineSnapshot(`
      Array [
        Object {
          "message": "{\\"text\\":\\"Hello World!\\"}",
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": "{\\"id\\":\\"john\\"}",
          "tag": "TRANSPORTATION_UPDATE",
        },
        Object {
          "message": "{\\"attachment\\":{\\"type\\":\\"image\\",\\"payload\\":{\\"url\\":\\"https://sociably.io/greeting.png\\"}},\\"quick_replies\\":[{\\"content_type\\":\\"text\\",\\"title\\":\\"Hi!\\",\\"payload\\":\\"ðŸ‘‹\\"}]}",
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": "{\\"id\\":\\"john\\"}",
          "tag": "TRANSPORTATION_UPDATE",
        },
      ]
    `);

    expect(apiStatus.isDone()).toBe(true);
  });
});

describe('#renderAttachment(message)', () => {
  const bot = new MessengerBot({ pageId, accessToken, appSecret });

  beforeEach(() => {
    bot.start();
  });

  afterEach(() => {
    bot.stop();
  });

  it('resolves null if message is empty', async () => {
    const empties = [undefined, null, [], <></>];
    for (const empty of empties) {
      await expect(bot.renderAttachment(empty)).resolves.toBe(null); // eslint-disable-line no-await-in-loop
    }
  });

  it('call message_attachments api', async () => {
    const apiStatus = graphApi.reply(200, [
      { code: 200, body: JSON.stringify({ attachment_id: 401759795 }) },
    ]);

    const response = await bot.renderAttachment(
      <Image url="https://sociably.io/trollface.png" />
    );
    expect(response).toMatchSnapshot();
    expect(response!.results).toEqual([
      {
        code: 200,
        body: { attachment_id: 401759795 },
      },
    ]);

    expect(bodySpy.mock).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    expect(body).toMatchSnapshot();
    const reqest = JSON.parse(body.batch)[0];
    expect(reqest.method).toBe('POST');
    expect(reqest.relative_url).toBe('me/message_attachments');
    expect(querystring.decode(reqest.body)).toMatchInlineSnapshot(`
      Object {
        "message": "{\\"attachment\\":{\\"type\\":\\"image\\",\\"payload\\":{\\"url\\":\\"https://sociably.io/trollface.png\\"}}}",
      }
    `);

    expect(apiStatus.isDone()).toBe(true);
  });
});

describe('#makeApiCall()', () => {
  it('call facebook graph api', async () => {
    const bot = new MessengerBot({ accessToken, pageId });
    bot.start();

    const apiCall = graphApi.reply(200, [{ code: 200, body: '{"foo":"bar"}' }]);

    await expect(
      bot.makeApiCall('POST', 'foo', { bar: 'baz' })
    ).resolves.toEqual({
      foo: 'bar',
    });

    expect(apiCall.isDone()).toBe(true);
  });

  it('throw MetaApiError if api call fail', async () => {
    const bot = new MessengerBot({ accessToken, pageId });
    bot.start();

    const apiCall = graphApi.reply(200, [
      {
        code: 400,
        body: JSON.stringify({
          error: {
            message: 'bad',
            type: 'devil',
            code: 444,
            error_subcode: 666,
            fbtrace_id: 'xxxx',
          },
        }),
      },
    ]);

    try {
      await bot.makeApiCall('POST', 'foo', { bar: 'baz' });
      expect('should not be here').toBeFalsy();
    } catch (err) {
      expect(err).toBeInstanceOf(MetaApiError);
      expect(err.code).toBe(444);
      expect(err.message).toBe('bad');
    }

    expect(apiCall.isDone()).toBe(true);
  });
});
