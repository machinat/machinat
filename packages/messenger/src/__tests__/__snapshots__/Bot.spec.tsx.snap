// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render(channel, message, options) render options 1`] = `
Object {
  "jobs": Array [
    Object {
      "assetTag": undefined,
      "fileData": undefined,
      "fileInfo": undefined,
      "key": "facebook.1234567890.john",
      "request": Object {
        "body": Object {
          "message": Object {
            "text": "Hello World!",
          },
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": Object {
            "id": "john",
          },
          "tag": "TRANSPORTATION_UPDATE",
        },
        "method": "POST",
        "relative_url": "me/messages",
      },
    },
    Object {
      "assetTag": undefined,
      "fileData": undefined,
      "fileInfo": undefined,
      "key": "facebook.1234567890.john",
      "request": Object {
        "body": Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": undefined,
                "is_reusable": undefined,
                "url": "https://sociably.io/greeting.png",
              },
              "type": "image",
            },
            "metadata": undefined,
            "quick_replies": Array [
              Object {
                "content_type": "text",
                "image_url": undefined,
                "payload": "ðŸ‘‹",
                "title": "Hi!",
              },
            ],
          },
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": Object {
            "id": "john",
          },
          "tag": "TRANSPORTATION_UPDATE",
        },
        "method": "POST",
        "relative_url": "me/messages",
      },
    },
  ],
  "results": Array [
    Object {
      "body": Object {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
    Object {
      "body": Object {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "assetTag": undefined,
          "fileData": undefined,
          "fileInfo": undefined,
          "key": "facebook.1234567890.john",
          "request": Object {
            "body": Object {
              "message": Object {
                "text": "Hello World!",
              },
              "messaging_type": "MESSAGE_TAG",
              "notification_type": "SILENT_PUSH",
              "persona_id": "billy17",
              "recipient": Object {
                "id": "john",
              },
              "tag": "TRANSPORTATION_UPDATE",
            },
            "method": "POST",
            "relative_url": "me/messages",
          },
        },
        Object {
          "assetTag": undefined,
          "fileData": undefined,
          "fileInfo": undefined,
          "key": "facebook.1234567890.john",
          "request": Object {
            "body": Object {
              "message": Object {
                "attachment": Object {
                  "payload": Object {
                    "attachment_id": undefined,
                    "is_reusable": undefined,
                    "url": "https://sociably.io/greeting.png",
                  },
                  "type": "image",
                },
                "metadata": undefined,
                "quick_replies": Array [
                  Object {
                    "content_type": "text",
                    "image_url": undefined,
                    "payload": "ðŸ‘‹",
                    "title": "Hi!",
                  },
                ],
              },
              "messaging_type": "MESSAGE_TAG",
              "notification_type": "SILENT_PUSH",
              "persona_id": "billy17",
              "recipient": Object {
                "id": "john",
              },
              "tag": "TRANSPORTATION_UPDATE",
            },
            "method": "POST",
            "relative_url": "me/messages",
          },
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`#render(channel, message, options) render options 2`] = `
Object {
  "access_token": "_ACCESS_TOKEN_",
  "appsecret_proof": "932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f",
  "batch": "[{\\"method\\":\\"POST\\",\\"relative_url\\":\\"me/messages\\",\\"body\\":\\"message=%7B%22text%22%3A%22Hello%20World!%22%7D&messaging_type=MESSAGE_TAG&tag=TRANSPORTATION_UPDATE&notification_type=SILENT_PUSH&persona_id=billy17&recipient=%7B%22id%22%3A%22john%22%7D\\",\\"omit_response_on_success\\":false,\\"name\\":\\"facebook.1234567890.john-1\\"},{\\"method\\":\\"POST\\",\\"relative_url\\":\\"me/messages\\",\\"body\\":\\"message=%7B%22attachment%22%3A%7B%22type%22%3A%22image%22%2C%22payload%22%3A%7B%22url%22%3A%22https%3A%2F%2Fsociably.io%2Fgreeting.png%22%7D%7D%2C%22quick_replies%22%3A%5B%7B%22content_type%22%3A%22text%22%2C%22title%22%3A%22Hi!%22%2C%22payload%22%3A%22%F0%9F%91%8B%22%7D%5D%7D&messaging_type=MESSAGE_TAG&tag=TRANSPORTATION_UPDATE&notification_type=SILENT_PUSH&persona_id=billy17&recipient=%7B%22id%22%3A%22john%22%7D\\",\\"omit_response_on_success\\":false,\\"depends_on\\":\\"facebook.1234567890.john-1\\",\\"name\\":\\"facebook.1234567890.john-2\\"}]",
}
`;

exports[`#render(channel, message, options) send messages to me/messages api 1`] = `
Object {
  "jobs": Array [
    Object {
      "assetTag": undefined,
      "fileData": undefined,
      "fileInfo": undefined,
      "key": "facebook.1234567890.john",
      "request": Object {
        "body": Object {
          "message": Object {
            "text": "Hello World!",
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "recipient": Object {
            "id": "john",
          },
          "tag": undefined,
        },
        "method": "POST",
        "relative_url": "me/messages",
      },
    },
    Object {
      "assetTag": undefined,
      "fileData": undefined,
      "fileInfo": undefined,
      "key": "facebook.1234567890.john",
      "request": Object {
        "body": Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": undefined,
                "is_reusable": undefined,
                "url": "https://sociably.io/greeting.png",
              },
              "type": "image",
            },
            "metadata": undefined,
            "quick_replies": Array [
              Object {
                "content_type": "text",
                "image_url": undefined,
                "payload": "ðŸ‘‹",
                "title": "Hi!",
              },
            ],
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "recipient": Object {
            "id": "john",
          },
          "tag": undefined,
        },
        "method": "POST",
        "relative_url": "me/messages",
      },
    },
  ],
  "results": Array [
    Object {
      "body": Object {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
    Object {
      "body": Object {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "assetTag": undefined,
          "fileData": undefined,
          "fileInfo": undefined,
          "key": "facebook.1234567890.john",
          "request": Object {
            "body": Object {
              "message": Object {
                "text": "Hello World!",
              },
              "messaging_type": undefined,
              "notification_type": undefined,
              "persona_id": undefined,
              "recipient": Object {
                "id": "john",
              },
              "tag": undefined,
            },
            "method": "POST",
            "relative_url": "me/messages",
          },
        },
        Object {
          "assetTag": undefined,
          "fileData": undefined,
          "fileInfo": undefined,
          "key": "facebook.1234567890.john",
          "request": Object {
            "body": Object {
              "message": Object {
                "attachment": Object {
                  "payload": Object {
                    "attachment_id": undefined,
                    "is_reusable": undefined,
                    "url": "https://sociably.io/greeting.png",
                  },
                  "type": "image",
                },
                "metadata": undefined,
                "quick_replies": Array [
                  Object {
                    "content_type": "text",
                    "image_url": undefined,
                    "payload": "ðŸ‘‹",
                    "title": "Hi!",
                  },
                ],
              },
              "messaging_type": undefined,
              "notification_type": undefined,
              "persona_id": undefined,
              "recipient": Object {
                "id": "john",
              },
              "tag": undefined,
            },
            "method": "POST",
            "relative_url": "me/messages",
          },
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`#render(channel, message, options) send messages to me/messages api 2`] = `
Object {
  "access_token": "_ACCESS_TOKEN_",
  "appsecret_proof": "932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f",
  "batch": "[{\\"method\\":\\"POST\\",\\"relative_url\\":\\"me/messages\\",\\"body\\":\\"message=%7B%22text%22%3A%22Hello%20World!%22%7D&recipient=%7B%22id%22%3A%22john%22%7D\\",\\"omit_response_on_success\\":false,\\"name\\":\\"facebook.1234567890.john-1\\"},{\\"method\\":\\"POST\\",\\"relative_url\\":\\"me/messages\\",\\"body\\":\\"message=%7B%22attachment%22%3A%7B%22type%22%3A%22image%22%2C%22payload%22%3A%7B%22url%22%3A%22https%3A%2F%2Fsociably.io%2Fgreeting.png%22%7D%7D%2C%22quick_replies%22%3A%5B%7B%22content_type%22%3A%22text%22%2C%22title%22%3A%22Hi!%22%2C%22payload%22%3A%22%F0%9F%91%8B%22%7D%5D%7D&recipient=%7B%22id%22%3A%22john%22%7D\\",\\"omit_response_on_success\\":false,\\"depends_on\\":\\"facebook.1234567890.john-1\\",\\"name\\":\\"facebook.1234567890.john-2\\"}]",
}
`;

exports[`#renderAttachment(message) call message_attachments api 1`] = `
Object {
  "jobs": Array [
    Object {
      "fileData": undefined,
      "fileInfo": undefined,
      "request": Object {
        "body": Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": undefined,
                "is_reusable": undefined,
                "url": "https://sociably.io/trollface.png",
              },
              "type": "image",
            },
          },
        },
        "method": "POST",
        "relative_url": "me/message_attachments",
      },
    },
  ],
  "results": Array [
    Object {
      "body": Object {
        "attachment_id": 401759795,
      },
      "code": 200,
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "fileData": undefined,
          "fileInfo": undefined,
          "request": Object {
            "body": Object {
              "message": Object {
                "attachment": Object {
                  "payload": Object {
                    "attachment_id": undefined,
                    "is_reusable": undefined,
                    "url": "https://sociably.io/trollface.png",
                  },
                  "type": "image",
                },
              },
            },
            "method": "POST",
            "relative_url": "me/message_attachments",
          },
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`#renderAttachment(message) call message_attachments api 2`] = `
Object {
  "access_token": "_ACCESS_TOKEN_",
  "appsecret_proof": "932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f",
  "batch": "[{\\"method\\":\\"POST\\",\\"relative_url\\":\\"me/message_attachments\\",\\"body\\":\\"message=%7B%22attachment%22%3A%7B%22type%22%3A%22image%22%2C%22payload%22%3A%7B%22url%22%3A%22https%3A%2F%2Fsociably.io%2Ftrollface.png%22%7D%7D%7D\\",\\"omit_response_on_success\\":false}]",
}
`;
